### PR TITLE
Wikidata <> authors integration: first steps

### PR DESCRIPTION
**Pull request by RayBB**
CLOSED IN FAVOR OF #9130

[Document with the goal of this project](https://docs.google.com/document/d/1-xAija9Pfhtwc-wCAgERHBx6GvFv40Hb-nfHd9f6SPc/edit?usp=drivesdk)

v0 goals:
- to start pulling wikidata into the database and showing something basic
- Upstream code class Author —> Add get_wikidata
- Should Fetch from wikidata based on parameter. Otherwise only check wikidata_cache table
- Author edit/view templates will call this
  - Render_wikidata_infobox method for rending the wikidata author box

Next steps:
- Decide on a bulk import method
- Enrich the infobox
- Use wikidata for author autocomplete

### Technical

Trying to keep it as simple as possible.
The wikidata method is on the authors model because works/editions store wikidata IDs differently so we'll have to handle that when we get there. My current coals is just authors.

### Testing

Add a wikidata ID to an author and then you'll start seeing this "short description" field showing up on the side.

### Screenshot

<details>
  <summary>Aug 30 demo video (slightly outdated), see below</summary>
  

https://github.com/internetarchive/openlibrary/assets/921217/509cdf8c-770d-4836-a8e2-631f468b32e3

  
  
</details>

<details>
  <summary>September 2 screenshot (most recent)</summary>
  
  
image

  
</details>

### Stakeholders

@cdrini

### CODE REVIEW

**Comment by davidscotson** [comment]
It would be really neat if this was added to the autocomplete, which currently pulls in the author's name, date-of-birth/death, genres and top work.

On the Wikidata side this is mostly designed to be used to differentiate between two identically named items, so it's an ideal use case.

It might also be worth looking at the Wikidata-powered infoboxes on Wikipedia to see what kind of info they surface.

https://en.wikipedia.org/wiki/Template:Infobox_writer/Wikidata

File: openlibrary/core/models.py
Comment (cclauss) [inline openlibrary/core/models.py:728]: [Memcached](https://en.wikipedia.org/wiki/Memcached) running on our hosts `ol-mem0`, ol-mem1`, and `ol-mem2` make remembering easy...

```suggestion
# Move this first import line up to the top of the file with the other imports!!
from openlibrary.core import cache

    @cache.memoize(engine="memcache", key="wikidata", expires=60 * 60)
    def wikidata(self, QID: str = "Q44"):
    """
    To accelerate repeated wikidata lookups, we use openlibrary.core.cache to store results on our memcache
    servers.  We keep these results in the cache for one hour (60 minutes * 60 seconds) before they expire.
    """
```

File: openlibrary/plugins/wikidata/code.py
Comment (cclauss) [inline openlibrary/plugins/wikidata/code.py:23]:
```suggestion
    url = f"https://www.wikidata.org/w/rest.php/wikibase/v0/entities/items/{QID}"
    response = requests.get(url).json()
```

File: openlibrary/plugins/wikidata/code.py
Comment (cclauss) [inline openlibrary/plugins/wikidata/code.py:32]:
```suggestion
    if entity := WikidataEntities.get_by_id(QID):
        return WikiDataEntity(descriptions=entity["descriptions"])

    response = requests.get(
        'https://www.wikidata.org/w/rest.php/wikibase/v0/entities/items/{QID}'
    )
    response.raise_for_status()
    wikdata = response.json()
    # TODO check for 200?
    WikidataEntities.add(QID, wikdata)
    return WikiDataEntity(descriptions=wikdata["descriptions"])
```

File: openlibrary/core/wikidata.py
Comment (cclauss) [inline openlibrary/core/wikidata.py:13]:
```suggestion
        if result := cls.get_by_ids([id]):
            return result[0]
        return None
```

File: openlibrary/core/models.py
Comment (cclauss) [inline openlibrary/core/models.py:8]: Given that we only support Python 3.11, my strong preference would be that we adhere to `ruff rule UP007` except in code that is directly related to Solr.  PEP604 makes code easier to read and eliminates noisy imports.
* https://beta.ruff.rs/docs/rules/non-pep604-annotation

% `ruff rule UP007`
# non-pep604-annotation (UP007)

Derived from the **pyupgrade** linter.

Autofix is sometimes available.

## What it does
Check for type annotations that can be rewritten based on [PEP 604] syntax.

## Why is this bad?
[PEP 604] introduced a new syntax for union type annotations based on the
`|` operator. This syntax is more concise and readable than the previous
`typing.Union` and `typing.Optional` syntaxes.

This rule is enabled when targeting Python 3.10 or later (see:
[`target-version`]). By default, it's _also_ enabled for earlier Python
versions if `from __future__ import annotations` is present, as
`__future__` annotations are not evaluated at runtime. If your code relies
on runtime type annotations (either directly or via a library like
Pydantic), you can disable this behavior for Python versions prior to 3.10
by setting [`pyupgrade.keep-runtime-typing`] to `true`.

## Example
```python
from typing import Union

foo: Union[int, str] = 1
```

Use instead:
```python
foo: int | str = 1
```

## Options
- `target-version`
- `pyupgrade.keep-runtime-typing`

[PEP 604]: https://peps.python.org/pep-0604/

Comment (RayBB) [reply openlibrary/core/models.py:8]: @cclauss why don't we want this rule for solr?
It seems the codebase is kinda split about as many uses of `Optional` as there are `| None` 

I also see you turned off this rule in our config, is that just because of solr?
https://github.com/internetarchive/openlibrary/blob/c7c1d52c395ced1dcb0e98eb5409ac4d85d2f137/pyproject.toml#L70

Comment (cclauss) [reply openlibrary/core/models.py:8]: `openlibrary/solr/types_generator.py` I believe still requires pre-PEP604 types.

**Comment by hornc** [comment]
What is the usecase behind this? Is there to be a bulk QID fetching UI or something?
Individual author lookups on an individual UI could be done with links or simple requests with a review feature.
I'm not sure what a Wikidata table is for.

**Comment by RayBB** [comment]
@hornc 
use case: this is an MVP to start integrating data from wikidata.
There will be many next steps to improving the reader experience and using the data for search.
Also I made this to lay out some ideas:
https://github.com/internetarchive/openlibrary/wiki/Wikidata-Integration

File: openlibrary/core/models.py
Comment (cdrini) [reply openlibrary/core/models.py]: I believe this was due to an issue with the FnToCLI ? But I updated that a while back ` | None` works. So whichever!

File: openlibrary/templates/type/author/edit.html
Comment (cdrini) [inline openlibrary/templates/type/author/edit.html:26]: Not the best place for it, but maybe toss it in olform under `.olform#addAuthor` for now.

File: openlibrary/templates/wikidata_author.html
Comment (cdrini) [inline openlibrary/templates/wikidata_author.html:10]:
```suggestion
        <h3 style="margin-top: 2px;margin-bottom: 2px;" class="bold black">$_('Quick Info')</h3>
```

File: openlibrary/templates/wikidata_author.html
Comment (cdrini) [inline openlibrary/templates/wikidata_author.html:17]: This will be kind of annoying to i18n ; I'd make the whole thing a link

```suggestion
            <a href="http://www.wikidata.org/entity/$wd_entity.id" style="font-size: .6em;">$_('Powered by Wikidata')
                <img src="/static/images/icons/edit.svg">
            </a>
```

File: openlibrary/templates/wikidata_author.html
Comment (cdrini) [inline openlibrary/templates/wikidata_author.html:1]: Put the css for this in a new less in `static/css/wikidatabox.less`.

Then import that CSS from `page-user.less` .

(I'm sorry, this flow is not great 😅 )

File: openlibrary/core/wikidata.py
Comment (cdrini) [reply openlibrary/core/wikidata.py:29]: Doesn't this need to be converted to `WikidataRow`...

File: openlibrary/plugins/wikidata/code.py
Comment (cdrini) [inline openlibrary/plugins/wikidata/code.py:32]: We can probably inline this function, and the `timedelta` helpers might be a little clearer here!

```suggestion
    if entity and (datetime.now() - entity.updated) < timedelta(days=30):
```

**Review by cdrini — COMMENTED** [review]
Nice! A few small things, but I think the main things are:

The split between the two files is a little confusing, especially since we have `WikidataEntities`, `WikidataEntity`, `WikidataRow` and the distinction between these is a little confusing. I think we want:

```
@dataclass
class WikidataEntity:
    id: str
    data: dict
    updated: datetime

    @property
    def age(self) -> timespan

class WikidataPlugin:
    def get_entity(id: str, cache_only = True) -> WikidataEntity | None
    
    def _get_from_web(id: str) -> WikidataEntity
    def _get_from_cache(id: str) -> WikidataEntity | None
    def _add_to_cache(id: str) -> None
```

That will I think keep all the code easily in one place, and since I doubt our wikidata caching will ever really grow to be more complicated than these, I think it'll keep things a touch tidier!

Actually since `WikidataPlugin` doesn't have any state it needs to keep track of, we can dissolve that class and just have those function in the module itself!

**Comment by RayBB** [comment]
@cdrini I think I've addressed all your concerns and the code is looking a lot cleaner.
Ready for your next round of feedback!

File: openlibrary/core/wikidata.py
Comment (cdrini) [inline openlibrary/core/wikidata.py:70]: This is only ever set to 30 or 0. We should simplify this to `allow_cache = True` or something. `ttl` is a bit of a technical term which isn't really relevant outside the users of this function. You can put the ttl as a constant at the top of this file ; `WIKIDATA_CACHE_TTL_DAYS = 30`. Or `WIKIDATA_CACHE_TTL = timedelta(days=30)` might be better to use the language native time features!

File: openlibrary/core/wikidata.py
Comment (cdrini) [inline openlibrary/core/wikidata.py:46]: Let's see if we can combine these two classes ; they only differ by one field, the `updated` field. We might want to rename this to `fetch_time` because then it makes sense in both contexts.

Right now the two very similarly named function are a little dissonant in terms of how close conceptually they are to each other.

**Review by cdrini — COMMENTED** [review]
This looks good to me! Would be great if we can merge those two classes + one other comment.

Next step for me is to create this table on prod so we can deploy this.

Next step for the project is to begin working on bulk import. We can make tweaks to how/where the data is displayed at any point, but getting the bulk data import sorted is likely the next most impactful step.

File: openlibrary/core/wikidata.py
Comment (cclauss) [reply openlibrary/core/wikidata.py:70]: https://en.wikipedia.org/wiki/Time_to_live ;-)

Comment (RayBB) [reply openlibrary/core/wikidata.py]: Fixed

File: openlibrary/core/wikidata.py
Comment (cdrini) [reply openlibrary/core/wikidata.py:53]: 2 would be preferable :+1: . `updated` works for me!

**Comment by cdrini** [comment]
Oh sorry if that was ambiguous I mean Option 2 is preferable, not 2 classes is preferable!

**Comment by tfmorris** [comment]
To reiterate [my comment](https://github.com/internetarchive/openlibrary/pull/8236#issuecomment-1697752558) from August, which no one apparently read, this is NOT an appropriate use of the Wikidata API. They make data dumps available for just such purposes as this. As the recent victim of, perhaps unintentional DOS events, I would hope that OpenLibrary would be sensitive to the load that they place on other portions of the open data ecosystem.

Additionally, the lookup should be done using the OL ID in the Wikidata dump, rather than QID in OpenLibrary, because there are a large number of entries which exist in Wikidata, but aren't linked from OpenLibrary.

**Comment by cdrini** [comment]
Ah whoops forgot to respond. This isn't a DDOS risk since it only makes a request on user _action_ ; going to the author edit page. (This page is also only accessible to logged in users).

The next stop of this project is 100% to bulk populate the DB from a wikidata dump :+1: 

>  there are a large number of entries which exist in Wikidata, but aren't linked from OpenLibrary.

Do you mean there are records in Wikidata with an open library id, but for which the open library record does not have a wikidata id? I don't _think_ this is a large dataset; @hornc 's bots should be syncing this up I _believe_? Not entirely sure he would know best on that one.

**Comment by RayBB** [comment]
> Oh sorry if that was ambiguous I mean Option 2 is preferable, not 2 classes is preferable! I wouldn't entirely remove the `data` field though, leave that in since it can be useful.

@cdrini can you elaborate on what you mean by not removing the data field?
Do you mean that the class should not have the data field but in the DB we should keep them all in the data field?
I can see that making sense since the fields (labels, etc) are unlikely to be queried directly by postgres.

File: openlibrary/core/wikidata.py
Comment (RayBB) [reply openlibrary/core/wikidata.py:53]: done, and it looks great :)

**Comment by hornc** [comment]
Responding because my name was mentioned again.

I still don't understand this PR / Wikidata project at all.

The existence of a Wikipedia database in Open Library _at all_ seems like a giant red flag to me, but from reading the available docs I can't point to anything specific.

From the first 2 lines of this PR
 > v0 goals:
>    to start pulling wikidata into the database and showing something basic

@tfmorris 's concerns about DDOS seem valid, I don't think the doc says anything about the scale. 

@cdrini 's comment about "only on user request" seems to mitigate that risk, but left me thinking what's the point of only storing a partial mirror of Wikidata on OL? Is it for web caching popular author pages responses? Maybe you should consider using web caching technology to achieve the same goal (please don't take that a serious suggestion, because I don't think that's the problem here).

The doc doesn't really give any indication why we need a database. From reading the linked doc https://github.com/internetarchive/openlibrary/wiki/Wikidata-Integration  the most concrete usecase and functional suggestion I could extract from it was the UI change to the Open Library author pages where an Open Library patron might want to see the corresponding data from Wikidata, side by side.

I'd say the core usecase covered by this "As an Open Library patron looking at a Open Library author page, I might be curious as to how this author is represented on a different website: Wikidata" is 100% satisfied by having the Wikidata link as part of our author metadata, so no action is required.

But from @cdrini's recent comment, it looks like it's only meant to be on the _edit_ page, so only librarians are meant to see this. I had a whole set of thoughts based on reading the document on how it was a bit pointless displaying a different variant to all patrons. It makes a little more sense if the use case is now revealed to be for helping librarians make metadata improvements, but that info wasn't available when this PR started, or when the document was created, so how could we constructively evaluate it against that purpose?

Anyway, based on my original reading of the doc and the recent revealed details:

* There seems no point doing a fragmentary mirror of Wikidata (apparently that is not being suggested, but it wasn't clear)
* It's apparently obvious that doing a full network fetch of Wikidata is a bad idea (that wasn't clear)
* The sensible way to do a full Wikidata db population into a SQL db on a different website with a different purpose is via db dumps... (nothing in the doc says or implies we need it or that's how it was going to be done)
.....
* Then what is the point of fetching _any_ data live on user request?

You'll probably tell me that it's obviously to get the latest data and keep it fresh, in which case what is the point of storing any of it in the database at all.... and why is that even important?

The point is we'll have to have a silly argument about databases when none of this stuff is at all obvious from the docs I was asked to read, and _requires_ a real time discussion to make any sense of it, and even then nothing is clear as to how any of it relates to improving anything to with Wikidata / OL integration, or even is the best way to achieve the minor but concrete UI changes that might (or might not!) be driving all of this development.

After reading the doc and establishing the only concrete suggestion were some small UI changes, I had specific concerns that the changes wouldn't be useful as there are only a few possibilities:

* The metadata is exactly the same
* The metadata is wildly different
* The metadata differs in trivial ways 

The doc gave no indication of what might be done in each case. I don't think it's at all obvious, other than in many cases it'll be "do nothing", and in others (perhaps considerably fewer) the appropriate action will be very specific and depend on other factors not even touched on here.

I also had some concerns about practicalities and licensing of the (reading between the lines using my knowledge of the two systems) the only two shared datafields that are likely to have any impact on OL -- images and descriptions....

but why would I bring _any_ of this up on a PR clearly about database implementation? Talking about UIs on a database PR is completely off topic....  And when the UI suggestion is sort of hedged as "maybe" example of what this integration will bring as we're only at "stage 0" I'm going to be jumping the gun to be talking about such details... Turns out I was slightly wrong anyway as it's apparently an edit page feature for a smaller group of users anyway, which puts a different spin on things.

Sorry if I'm sounding negative, but I am feeling quite a bit of frustration with these recent "stage 0" general efforts do do "something" leading with a big technical implementation which promises to change literally nothing at first (seemingly to counter any possible criticism) to open a whole range of future possibilities that we'll cover later. There's temptation to review in good faith a provide helpful suggestions, and then at least check that there are no unintended consequences, but most of that comes down to mechanical line length checking and whether the variables are clearly named.

The docs that exist are all about maybe possibilities and there's so much ambiguity it's impossible to offer any concrete evaluation of a "solution". I'm not just talking about this now, I had similar issues with the subject tags project, which again seemed like possibly relatively minor UI changes driving large code changes, but I had even less context to offer any comment, and the whole thing seemed designed to avoid any good faith critical interaction with it.

Now I don't know if I'm supposed to be talking about databases (I really don't want to be), efficient use of network requests, or specific UI details, but have the strong impression that anything I want to talk about driven by a desire to improve  Wikidata / OL integration can easily now be dismissed as off-topic noise on this PR and project, and that's the most frustrating thing about this is that now presumably multiple people clearly sharing the same goal to make better use of Wikidata are going to be reduced to arguing about irrelevant details.

**The structure of these implementation-first efforts are not conducive to canvasing constructive feedback. The use of the term "Stage 0" appears to be a indicator of this sort of thing.**

@tfmorris suggestions about going the other way, from Wikidata to OL seem more valuable to me, I'd consider a browser plugin to help librarians edit OL author pages if that's what we're trying to do... I also have thoughts about what you might expect to see given what I know about how OL got Wikidata QIDs (my historical bot matching, the Wikidata Mix'n'Match tool, and the actions of Wikidata editors)... but all these things seem explicitly not relevant to what's being discussed here.

Some concrete suggestions / alternatives which are on-topic for this PR:

* Try-it-and-see db mirrors of all of Wikidata are probably better suited to stand-alone projects and not part of OL's core codebase (in my opinion)
* Come up with a better post-hoc reason why OL needs a database named Wikidata
* Consider addressing Librarian UI editing issues directly, what OL  usecases require particular Author metadata and how these are not met currently, and how Wikidata might help by going back to planning stages and figuring out what will be impactful, and what is available. 

Apologies again for sounding negative, but I keep getting pinged on this and I would like to help make OL / Wikidata better, but the way this is set up I feel trapped, my other option would be to err on the side of just not saying anything (like I did for subject tags). I know I don't need to understand everything, but I'm also not going to review all code just because it exists. My problem here is really with the implementation-first trend, and while I'd like to talk about Wikidata, I'm not even sure where to begin here.

**Comment by cdrini** [comment]
Hey Charles! I've read through your feedback and am thinking on how to best respond. But regardless I'm sorry the process has made you feel frustrated; not at all the intention! Thank you for taking the time to write your thoughts.

**Comment by davidscotson** [comment]
I'd be interested in hearing more about what @hornc has to say about this bit:

> I also have thoughts about what you might expect to see given what I know about how OL got Wikidata QIDs (my historical bot matching, the Wikidata Mix'n'Match tool, and the actions of Wikidata editors)

Is #710 the right place to be discussing how to leverage wikidata in Open Library (and vice versa) in general?

**Comment by cdrini** [comment]
Hey! Ok @RayBB and I just spent an hour working through your feedback and trying to concisely hammer down some of our thoughts about this project!

We've come up with a new document for the project roadmap/plan here, that should hopefully provide extra context on how the steps in this PR relate to the longer term vision, and what the longer term vision is: https://docs.google.com/document/d/1-xAija9Pfhtwc-wCAgERHBx6GvFv40Hb-nfHd9f6SPc/edit

To address some of the direct comments here:

1. Loading a wikidata data dump is part of the plan, it is V1
2. This PR implements some of the steps which will be required by the next phases of the project, while also attempting to provide some of the value up-front by displaying the wikidata "infocard" on author pages
3. W.r.t your concerns about that "stage 0" leading nowhere, I think here, this stage 0 unlocks essential pieces of the next stages, and also provides direct value to patrons, while also exposing a piece of the UI which we want more feedback on.
4. For the DDOS concerns, yes completely valid! Although Ray has gotten confirmation from Wikidata folks that the volume might be ok, we do want to follow best practices here, which is why we will _not_ fetch wikidata on page view until we complete v1, the bulk loading of the wikidata database dump.
5. We'll also get feedback from the design folks on the wikidata infobox to determine what/how we want to show.

Ok, I hope that addresses most of the concerns here! Thanks for your patience (especially @RayBB  :P) Personally I think aligning ourselves with the open data universe is always net good, and having ways to easily fetch/integrate Wikidata data will open the doors for unique visualizations/queries which no other site can perform. It will also potentially help us evolve our own platform, if we find things we love about how Wikidata works!

**Comment by mekarpeles** [comment]
@RayBB we may need a rebase on this one and perhaps a reminder if there's any next steps you also may need from us! Thank you!!

**Comment by RayBB** [comment]
@mekarpeles I still need to change a few things like make sure it only fetches when viewing the edit page.

It isn't blocking this PR but I'd love some input from the design team on how they think we should make the infobox look when we add Sex, Date/place of birth/death, and more. We could keep something similar to the wikipedia infobox to start but I'm open to ideas.

**Comment by RayBB** [comment]
Blocked on #8949 - after this is merged I'll:
1. Rework the UI component.
2. Update how the caching works.

File: openlibrary/core/wikidata.py
Comment (cdrini) [inline openlibrary/core/wikidata.py:75]:
```suggestion
    qid: str, bust_cache: bool = False, fetch_missing: bool = False
```

File: openlibrary/core/wikidata.py
Comment (cdrini) [inline openlibrary/core/wikidata.py:29]: It seems like this is the type for these language-y things

```suggestion
    labels: dict[str, str]
```

File: openlibrary/core/wikidata.py
Comment (cdrini) [inline openlibrary/core/wikidata.py:34]: Because this is the only field that is not actually returned by https://www.wikidata.org/w/rest.php/wikibase/v0/entities/items/Q42 , let's label it as such so that it's clear this is internal to our application

```suggestion
    _updated: datetime  # This is when we fetched the data, not when the entity was changed in Wikidata
```

File: openlibrary/core/wikidata.py
Comment (cdrini) [inline openlibrary/core/wikidata.py:36]: Let's call these get_BLAH ; besides being more clear that these are functions, it also reduces the likelihood of types between this and `descriptions` which differ only by a letter.

```suggestion
    def get_description(self, language: str = 'en') -> str | None:
```

File: openlibrary/core/wikidata.py
Comment (cdrini) [inline openlibrary/core/wikidata.py:53]: We can rename this to:

```suggestion
    def to_json(self) -> str:
```

or we might not even need a method for this, and just use `json.dumps(wikidata_entity.asdict())`. It'll include the `_updated` field in postgres, but that's fine.

File: openlibrary/core/wikidata.py
Comment (cdrini) [inline openlibrary/core/wikidata.py:50]: This should work since we're taking all the fields exactly.

```suggestion
        return cls(
            **response,
            _updated: updated,
        )
```

File: openlibrary/core/wikidata.py
Comment (cdrini) [inline openlibrary/core/wikidata.py:106]: We unfortunately don't have a best practices for handling cases like this, but adding a logger.error should let us see if these are happening in Sentry.

```suggestion
        logger.error(f'Wikidata Response: {response.status_code}`, data={'id': id})
        return None
```

File: openlibrary/core/wikidata.py
Comment (cdrini) [inline openlibrary/core/wikidata.py:127]: Since the method returns a list, an empty list is falsy so we don't need the `len` check.

```suggestion
    if result := _get_from_cache_by_ids([id]):
```

File: openlibrary/plugins/wikidata/code.py
Comment (cdrini) [inline openlibrary/plugins/wikidata/code.py:1]: Assume this should be deleted.

**Review by cdrini — CHANGES_REQUESTED** [review]
Nice! Logic looks good; a few style things. I created the `wikidata` table and will toss this up on testing now since the logic is already :+1:

File: openlibrary/core/wikidata.py
Comment (RayBB) [reply openlibrary/core/wikidata.py:106]: I didn't see any examples in the code adding data like that so I put it in a string for now.

File: openlibrary/core/wikidata.py
Comment (RayBB) [reply openlibrary/core/wikidata.py]: If we try to do `json.dumps(asdict(entity))` then we get `Object of type datetime is not JSON serializable`.

However, I did rename it to something that seems clearer: `to_wikidata_api_json_format`

File: openlibrary/templates/type/author/view.html
Comment (RayBB) [inline openlibrary/templates/type/author/view.html:182]: I am leaving this code here now for testing.openlibrary.org to be easier to verify this PR on.

I can either remove it before we merge this or there is already a commit to remove it in This is removed in https://github.com/internetarchive/openlibrary/pull/9130

**Comment by RayBB** [comment]
Closing in favor of #9130

**Pull request closed without merging**
